doc/man/man1/normtiff.1
.\" RCSid $Id: normtiff.1,v 1.8 2024/10/02 15:58:56 greg Exp $
.TH NORMTIFF 1 2/25/99 RADIANCE
.SH NAME
normtiff - tone-map and convert RADIANCE picture or HDR TIFF to standard TIFF
.SH SYNOPSIS
.B normtiff
[
.B options
]
.B input
.B output.tif
.SH DESCRIPTION
.I Normtiff
prepares a Radiance picture or high dynamic-range TIFF
for output to a display or hard copy device.
If the dynamic range of the scene exceeds that of the display (as is
usually the case),
.I normtiff
will compress the dynamic range of the picture such that both
dark and bright regions are visible.
In addition, certain limitations in human vision may be mimicked in
order to provide an appearance similar to the experience one might
have in the actual scene.
.PP
Output is always an uncompressed RGB TIFF, which must be named
on the command line along with the input file.
If the input file has a ".tif" or ".tiff" extension,
.I normtiff
attempts to read it as a TIFF.
Otherwise,
.I normtiff
first tries opening it as a RADIANCE picture, only opening it
as a TIFF if it fails header inspection.
(See the
.I getinfo(1)
program.)\0
If the input is neither a RADIANCE picture nor a high dynamic-range TIFF,
the program reports an error and exits.
Hyperspectral radiance pictures are silently converted to RGB, though
the colors will be more accurate if the HSR picture is passed through
.I ra_xyze(1),
first.
.PP
The following command line options are understood.
Since this program is very similar to
.I pcond(1),
several of the switches are identical.
.TP 10n
.BR -b
Toggle 8-bit black and white (grayscale) TIFF output.
If the input is a grayscale TIFF, this switch is
automatically selected.
Otherwise, the output defaults to 24-bit RGB.
.TP
.BR -z
Output LZW-compressed TIFF (smaller file).
.TP
.BR -h
Mimic human visual response in the output.
The goal of this process is to produce output that correlates
strongly with a person's subjective impression of a scene.
This switch turns on both the
.I \-s
and
.I \-c
switches, described below.
.TP
.BR -s
Toggle the use of the human contrast sensitivity function in determining the
exposure for the image.
A darker scene will have relatively lower exposure with lower
contrast than a well-lit scene.
.TP
.BR -c
Toggle mesopic color correction.
If parts of the image are in the mesopic or scotopic range where
the cone photoreceptors lose their efficiency, this switch will
cause a corresponding loss of color visibility in the output and a
shift to a scotopic (blue-dominant) response function.
.TP
.BR -l
Toggle the use of a linear response function versus the standard dynamic
range compression algorithm.
This may make some parts of the resulting image too
dark or too bright to see.
.TP
.BI -u \ Ldmax
Specifies the top of the luminance range for the target output device.
That is, the luminance (in candelas/m^2) for an output pixel value
of (R,G,B)=(255,255,255).
This parameter affects tone mapping only when the
.I \-s
switch is on.
The default value is 100 cd/m^2.
.TP
.BI -d \ Lddyn
Specifies the dynamic range for the target output device, which is
the ratio of the maximum and minimum usable display luminances.
The default value is 100, which is typical for LCD monitors.
.TP
.BI -p " xr yr xg yg xb yb xw yw"
Specifies the RGB primaries for the target output device.
These are the 1931 CIE (x,y) chromaticity values for red, green,
blue and white, respectively.
.TP
.BI -g \ gamma
Specifies the output device gamma correction value.
The default value is 2.2, which is appropriate for most CRT monitors.
(A value of 1.8 is common in color prepress and color printers.)\0
.SH EXAMPLES
To convert a RADIANCE picture to an 8-bit grayscale TIFF:
.IP "" .2i
normtiff \-b scene.hdr sceneb.tif
.PP
To condition a high dynamic-range TIFF for a particular film recorder with
known color primaries, dynamic range and gamma response:
.IP "" .2i
pcond \-d 50 \-g 2.5 \-p .580 .340 .281 .570 .153 .079 .333 .333 orig.tif filmrgb.tif
.PP
To simulate human visual response on a monitor with known maximum luminance:
.IP "" .2i
normtiff \-h \-u 80 scene.hdr sceneh.tif
.SH REFERENCE
Greg Ward Larson, Holly Rushmeier, Christine Piatko,
``A Visibility Matching Tone Reproduction Operator for High Dynamic Range
Scenes,''
.I "IEEE Transactions on Visualization and Computer Graphics",
December 1997.
.PP
http://positron.cs.berkeley.edu/gwlarson/pixformat/
.SH AUTHOR
Greg Ward
.SH "SEE ALSO"
getinfo(1), pcond(1), pflip(1), 
pvalue(1), protate(1), ra_xyze(1), rpict(1), ximage(1)


src/px/normtiff.c
#ifndef lint
static const char	RCSid[] = "$Id: normtiff.c,v 3.17 2024/10/02 15:58:56 greg Exp $";
#endif
/*
 * Tone map SGILOG TIFF or Radiance picture and output 24-bit RGB TIFF
 */

#include <math.h>

#include "rtio.h"
#include "platform.h"
#include "tiffio.h"
#include "color.h"
#include "tonemap.h"
#include "tmaptiff.h"
#include "resolu.h"


TIFF	*tifout;			/* TIFF output */
int	flags = TM_F_CAMERA;		/* tone-mapping flags */
RGBPRIMP	rgbp = stdprims;	/* display primaries */
RGBPRIMS	myprims;		/* overriding display primaries */
double	ldmax = 100.;			/* maximum display luminance */
double	lddyn = 100.;			/* display dynamic range */
double	gamv = 2.2;			/* display gamma value */

short	ortab[8] = {		/* orientation conversion table */
	YMAJOR|YDECR,
	YMAJOR|YDECR|XDECR,
	YMAJOR|XDECR,
	YMAJOR,
	YDECR,
	XDECR|YDECR,
	XDECR,
	0
};

typedef struct {
	FILE	*fp;		/* file pointer */
	char	fmt[MAXFMTLEN];	/* picture format */
	double	pa;		/* pixel aspect ratio */
	RESOLU	rs;		/* picture resolution */
} PICTURE;

uint16	comp = COMPRESSION_NONE;	/* TIFF compression mode */

#define closepicture(p)		(fclose((p)->fp),free(p))

static gethfunc headline;

static int headline(char *s, void *pp);
static PICTURE *openpicture(char *fname);
static int tmap_picture(char *fname, PICTURE *pp);
static int tmap_tiff(char *fname, TIFF *tp);
static int putimage(uint16 or, uint32 xs, uint32 ys, float xr, float yr,
	uint16 ru, uby8	*pd);


int
main(
	int	argc,
	char	*argv[]
)
{
	PICTURE	*pin = NULL;
	TIFF	*tin = NULL;
	int	i, rval;

	for (i = 1; i < argc && argv[i][0] == '-'; i++)
		switch (argv[i][1]) {
		case 'h':			/* human observer settings */
			flags = TM_F_HUMAN;
			break;
		case 's':			/* toggle human contrast */
			flags ^= TM_F_HCONTR;
			break;
		case 'c':			/* toggle mesopic sens. */
			flags ^= TM_F_MESOPIC;
			break;
		case 'l':			/* toggle linear mapping */
			flags ^= TM_F_LINEAR;
			break;
		case 'b':			/* toggle greyscale output */
			flags ^= TM_F_BW;
			break;
		case 'g':			/* set display gamma */
			if (argc-i < 2) goto userr;
			gamv = atof(argv[++i]);
			break;
		case 'u':			/* set display maximum */
			if (argc-i < 2) goto userr;
			ldmax = atof(argv[++i]);
			break;
		case 'd':			/* set display dynamic range */
			if (argc-i < 2) goto userr;
			lddyn = atof(argv[++i]);
			break;
		case 'z':			/* LZW compression */
			comp = COMPRESSION_LZW;
			break;
		case 'p':			/* set display primaries */
			if (argc-i < 9) goto userr;
			myprims[RED][CIEX] = atof(argv[++i]);
			myprims[RED][CIEY] = atof(argv[++i]);
			myprims[GRN][CIEX] = atof(argv[++i]);
			myprims[GRN][CIEY] = atof(argv[++i]);
			myprims[BLU][CIEX] = atof(argv[++i]);
			myprims[BLU][CIEY] = atof(argv[++i]);
			myprims[WHT][CIEX] = atof(argv[++i]);
			myprims[WHT][CIEY] = atof(argv[++i]);
			rgbp = myprims;
			break;
		default:
			goto userr;
		}
	if (argc-i < 2) goto userr;
	if ((pin = openpicture(argv[i])) == NULL &&
			(tin = TIFFOpen(argv[i], "r")) == NULL) {
		fprintf(stderr, "%s: cannot open or interpret file \"%s\"\n",
				argv[0], argv[i]);
		exit(1);
	}
	if ((tifout = TIFFOpen(argv[i+1], "w")) == NULL) {
		fprintf(stderr, "%s: cannot open output TIFF \"%s\"\n",
				argv[0], argv[i+1]);
		exit(1);
	}
	if (pin != NULL) {
		rval = tmap_picture(argv[i], pin);
		closepicture(pin);
	} else {
		rval = tmap_tiff(argv[i], tin);
		TIFFClose(tin);
	}
	TIFFClose(tifout);
	exit(rval==0 ? 0 : 1);
userr:
	fprintf(stderr,
"Usage: %s [-h][-s][-c][-l][-b][-g gv][-d ld][-u lm][-z][-p xr yr xg yg xb yb xw yw] input.{tif|hdr} output.tif\n",
			argv[0]);
	exit(1);
}


static int
headline(				/* process line from header */
	char	*s,
	void *pp
)
{
	char	*cp;

	for (cp = s; *cp; cp++)
		if (*cp & 0x80)
			return(-1);	/* non-ascii in header */
	if (isaspect(s))
		((PICTURE *)pp)->pa *= aspectval(s);
	else
		formatval(((PICTURE *)pp)->fmt, s);
	return(0);
}


static PICTURE *
openpicture(			/* open/check Radiance picture file */
	char	*fname
)
{
	FILE	*fp;
	PICTURE	*pp;
	char	*cp;
					/* check filename suffix */
	if (fname == NULL) return(NULL);
	for (cp = fname; *cp; cp++)
		;
	while (cp > fname && cp[-1] != '.')
		if (*--cp == '/') {
			cp = fname;
			break;
		}
	if (cp > fname && !strncasecmp(cp, "tif", 3))
		return(NULL);		/* assume it's a TIFF */
					/* else try opening it */
	if ((fp = fopen(fname, "r")) == NULL)
		return(NULL);
	SET_FILE_BINARY(fp);
					/* allocate struct */
	if ((pp = (PICTURE *)malloc(sizeof(PICTURE))) == NULL)
		return(NULL);		/* serious error -- should exit? */
	pp->fp = fp; pp->fmt[0] = '\0'; pp->pa = 1.;
					/* load header */
	if (getheader(fp, headline, pp) < 0) {
		closepicture(pp);
		return(NULL);
	}
	if (!pp->fmt[0])		/* assume RGBE if unspecified */
		strcpy(pp->fmt, COLRFMT);
	if ((!globmatch(PICFMT, pp->fmt) && strcmp(SPECFMT, pp->fmt))
			|| !fgetsresolu(&pp->rs, fp)) {
		closepicture(pp);	/* failed test -- close file */
		return(NULL);
	}
	rewind(fp);			/* passed test -- rewind file */
	return(pp);
}


static int
tmap_picture(			/* tone map Radiance picture */
	char	*fname,
	PICTURE	*pp
)
{
	uint16	orient;
	double	paspect = (pp->rs.rt & YMAJOR) ? pp->pa : 1./pp->pa;
	int	xsiz, ysiz;
	uby8	*pix;
					/* read and tone map picture */
	if (tmMapPicture(&pix, &xsiz, &ysiz, flags,
			rgbp, gamv, lddyn, ldmax, fname, pp->fp) != TM_E_OK)
		return(-1);
					/* figure out TIFF orientation */
	for (orient = 8; --orient; )
		if (ortab[orient] == pp->rs.rt)
			break;
	orient++;
					/* put out our image */
	if (putimage(orient, (uint32)xsiz, (uint32)ysiz,
			72., 72./paspect, 2, pix) != 0)
		return(-1);
					/* free data and we're done */
	free(pix);
	return(0);
}


static int
tmap_tiff(			/* tone map SGILOG TIFF */
	char	*fname,
	TIFF	*tp
)
{
	float	xres, yres;
	uint16	orient, resunit, phot;
	int	xsiz, ysiz;
	uby8	*pix;
					/* check to make sure it's SGILOG */
	TIFFGetFieldDefaulted(tp, TIFFTAG_PHOTOMETRIC, &phot);
	if ((phot == PHOTOMETRIC_LOGL) | (phot == PHOTOMETRIC_MINISBLACK))
		flags |= TM_F_BW;
					/* read and tone map TIFF */
	if (tmMapTIFF(&pix, &xsiz, &ysiz, flags,
			rgbp, gamv, lddyn, ldmax, fname, tp) != TM_E_OK)
		return(-1);
					/* get relevant tags */
	TIFFGetFieldDefaulted(tp, TIFFTAG_RESOLUTIONUNIT, &resunit);
	TIFFGetFieldDefaulted(tp, TIFFTAG_XRESOLUTION, &xres);
	TIFFGetFieldDefaulted(tp, TIFFTAG_YRESOLUTION, &yres);
	TIFFGetFieldDefaulted(tp, TIFFTAG_ORIENTATION, &orient);
					/* put out our image */
	if (putimage(orient, (uint32)xsiz, (uint32)ysiz,
			xres, yres, resunit, pix) != 0)
		return(-1);
					/* free data and we're done */
	free(pix);
	return(0);
}


static int
putimage(	/* write out our image */
	uint16	or,
	uint32	xs,
	uint32	ys,
	float	xr,
	float	yr,
	uint16	ru,
	uby8	*pd
)
{
	int	y;
	uint32	rowsperstrip;

	TIFFSetField(tifout, TIFFTAG_COMPRESSION, COMPRESSION_NONE);
	if (flags & TM_F_BW) {
		TIFFSetField(tifout, TIFFTAG_PHOTOMETRIC,
				PHOTOMETRIC_MINISBLACK);
		TIFFSetField(tifout, TIFFTAG_SAMPLESPERPIXEL, 1);
	} else {
		TIFFSetField(tifout, TIFFTAG_PHOTOMETRIC, PHOTOMETRIC_RGB);
		TIFFSetField(tifout, TIFFTAG_SAMPLESPERPIXEL, 3);
	}
	if (rgbp != stdprims) {
		TIFFSetField(tifout, TIFFTAG_PRIMARYCHROMATICITIES,
				(float *)rgbp);
		TIFFSetField(tifout, TIFFTAG_WHITEPOINT, (float *)rgbp[WHT]);
	}
	TIFFSetField(tifout, TIFFTAG_BITSPERSAMPLE, 8);
	TIFFSetField(tifout, TIFFTAG_IMAGEWIDTH, xs);
	TIFFSetField(tifout, TIFFTAG_IMAGELENGTH, ys);
	TIFFSetField(tifout, TIFFTAG_RESOLUTIONUNIT, ru);
	TIFFSetField(tifout, TIFFTAG_COMPRESSION, comp);
	TIFFSetField(tifout, TIFFTAG_XRESOLUTION, xr);
	TIFFSetField(tifout, TIFFTAG_YRESOLUTION, yr);
	TIFFSetField(tifout, TIFFTAG_PLANARCONFIG, PLANARCONFIG_CONTIG);
	TIFFSetField(tifout, TIFFTAG_ORIENTATION, or);
					/* compute good strip size */
	rowsperstrip = 8192/TIFFScanlineSize(tifout);
	if (rowsperstrip < 1) rowsperstrip = 1;
	TIFFSetField(tifout, TIFFTAG_ROWSPERSTRIP, rowsperstrip);
					/* write out scanlines */
	if (flags & TM_F_BW) {
		for (y = 0; y < ys; y++)
			if (TIFFWriteScanline(tifout, pd + y*xs, y, 0) < 0)
				goto writerr;
	} else {
		for (y = 0; y < ys; y++)
			if (TIFFWriteScanline(tifout, pd + y*3*xs, y, 0) < 0)
				goto writerr;
	}
	return(0);			/* all done! */
writerr:
	fputs("Error writing TIFF output\n", stderr);
	return(-1);
}
